qiskit/transpiler/passes/optimization/light_cone.py

```diff
@@ -0,0 +1,141 @@
+            same expectation value for the observable.
+        """
+        super().__init__()
+        self.observable = observable
+
+    @staticmethod
+    def _find_measurement_qubits(dag: DAGCircuit):
+        """For now this method assumes that all the measurements in the circuit are
+        final measurements."""
+        qubits_measured = set()
+        for node in dag.topological_nodes():
+            # print(getattr(node, "name", False))
+            if getattr(node, "name", False) == "measure":
+                qubits_measured |= {x for x in node.qargs}
+        print(qubits_measured)
+        return qubits_measured
+
+    def _get_initial_lightcone(
+        self, dag: DAGCircuit
+    ) -> tuple[set[int], tuple[Instruction, list[int]]]:
+        """Returns the initial lightcone.
+        If obervable is None, the lightcone is a different Z operator in each one of the
+        qubits with a measurement. If a Pauli observable is defined, that will be
+        the first gate in the circuit.
+        """
+        if self.observable is None:
+            qubits_measured = self._find_measurement_qubits(dag)
+            light_cone = qubits_measured
+            light_cone_ops = [(PauliGate("Z"), [qubit_index]) for qubit_index in qubits_measured]
+        else:
+            pauli_string = str(self.observable)
+            light_cone = [dag.qubits[i] for i, p in enumerate(pauli_string) if p != "I"]
+            light_cone_ops = [(PauliGate(pauli_string.replace("I", "")), light_cone)]
+
+        return set(light_cone), light_cone_ops
+
+    def run(self, dag: DAGCircuit):
+        """Run the LightCone pass on `dag`.
+
+        Args:
+            dag (DAGCircuit): the DAG to be optimized.
+
+        Returns:
+            DAGCircuit: the optimized DAG.
+        """
+        light_cone, light_cone_ops = self._get_initial_lightcone(dag)
+
+        print(light_cone, light_cone_ops)
+        # iterate over the nodes in reverse topological order
+        for node in reversed(list(dag.topological_op_nodes())):
+            if not light_cone.intersection(node.qargs):
+                dag.remove_op_node(node)
+                continue
+
+            commutes_bool = True
+            for op in light_cone_ops:
+                commute_bool = commutator.commute(op[0], op[1], [], node.op, node.qargs, [])
+                if not commute_bool:
+                    light_cone.update(node.qargs)
+                    light_cone_ops.append((node.op, node.qargs))
+                    commutes_bool = False
+                    break
+
+            if commutes_bool:
+                dag.remove_op_node(node)
+
+        return dag
+
+
+if __name__ == "__main__":
+    """This main will be only for testing purposes and will be removed
+    before merging the PR."""
+    from qiskit.circuit.library import RealAmplitudes
+    from qiskit.converters import circuit_to_dag
+
+    qc = RealAmplitudes(10, entanglement="pairwise").decompose()
```

Comment:
That's a good start for tests, you can just create a new file in `test/python/transpiler` that contains tests for this. Beyond testing that it works, we should also ensure failures are correctly handled, such as no measurements and no `observable`, an `observable` with mismatching size to the dag, checking mid-circuit measurements, ...

Reply:
Great!